python/ngen_cal/src/ngen/cal/__main__.py

```diff
@@ -62,6 +95,9 @@ def main(general: General, model_conf: "Mapping[str, Any]"):
         #    func(start_iteration, general.iterations, catchment_set, agent)
         func(start_iteration, general.iterations, agent)
 
+    # call `ngen_cal_finish` plugin hook functions
+    plugin_manager.hook.ngen_cal_finish()
```

Comment:
> should we guarantee that this is always called?

I find myself unsure also, though I don't know/remember what use cases you had in mind already when you put this together.  Might it be worth doing both by adding a separate _finally_-style hook?

Reply:
> I find myself unsure also, though I don't know/remember what use cases you had in mind already when you put this together. Might it be worth doing both by adding a separate _finally_-style hook?

The goal is to make something that is really flexible that more or less acts like an event system. I am hesitant to add more hooks as i'd like to keep the plugin surface as small as possible. I agree though, that having a finally style hook is probably the most useful here. I think the most sane thing is _always_ call `ngen_cal_finish` and pass any runtime exceptions to `ngen_cal_finish`. If a plugin wants to inspect that it can, but its not a requirement. Thoughts?

Reply:
The primary use case coming from the integration of calibration feature development was the ability to send e-mail notifications when a calibration job finished.  It would be probably be good in that use case to be able to notify about failures as well, so catching exceptions and passing them to the plugin to do something with is probably appropriate in that setting as well.

Reply:
Yeah, that was my thinking as well. Ill make the change! Thanks for the feedback, guys!